keps/sig-cluster-lifecycle/20190122-Certificates-copy-for-kubeadm-join--control-plane.md

```diff
@@ -0,0 +1,330 @@
+required for joining a new control plane node.
+In other words, it won't be possible to use kubeadm `--upload-certs` for copying a subset
+of certificates only.
+
+#### The TTL-token
+
+The TTL-token is a regular bootstrap token, with a short duration, but without
+any assigned usage or groups (see [Authenticating with Bootstrap Tokens](https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#bootstrap-token-secret-format)
+for more details).
+
+This token is created with the only purpose of triggering automatic deletion of
+the `kubeadm-certs` secret when deleted (the two objects will be linked via
+`ownerReference` attributes).
+
+The `TokenCleaner` controller will ensure automatic deletion of bootstrap tokens
+and therefore also deletion of the TTL-token.
+
+### Risks and Mitigations
+
+In case the cluster administrator uses the `kubeadm init --upload-certs` flag,
+the `kubeadm-certs` secret will store sensitive cluster certificates. Following
+security measures apply to the `kubeadm-certs` secret:
+
+- the `kubeadm-certs` secret is time lived (with a duration shorter than
+  bootstrap tokens)
+- if an attacker get access to the cluster when the `kubeadm-certs` secret still
+  exists, the cluster certs contained in the secret are protected by RBAC rules
+- if an attacker get access to the cluster when the `kubeadm-certs` secret still
+  exists with an identity that is authorized to read the `kubeadm-certs` secret,
+  the cluster certificates contained in the secret will be protected by the
+  encryption algorithm (without the encryption key, the secret contains a harmless
+  blog of bytes)
+
+The following security measures apply to the encryption key:
+
+- the encryption key will never be stored on the cluster by kubeadm
+- the encryption key will never be transmitted on the wire by kubeadm
+- the encryption key is implicitly time lived (because when the `kubeadm-certs`
+  secret is automatically deleted, the encryption key became useless)
+- it will be possible to hide the encryption key from the `kubeadm init` output
+  by using `--skip-token-print` or similar flag (but this requires the user
+  passing an encryption key to kubeadm using the kubeadm config file), thus
+  avoiding the key improperly registered in logs
+- it will be possible to use a config file for passing the encryption key to
+  `kubeadm join` command, thus avoiding the key improperly registered in
+  shell history
+
+Another possible risk derives from kubeadm allowing the user to fully customize
+control plane components, including also the possibility to remove the TokenCleaner
+from the list of enabled controllers in the kube-scheduler configuration, and as
+a consequence, disable the mechanisms that manages time lived tokens/time lived
+`kubeadm-certs` secret/time lived encryption key.
+
+A possible mitigation action for this risk is to enforce the TokenCleaner in case
+when the user opts-in in the upload-certs workflow defined in this document.
```

Comment:
enforcing might be the right thing to do.
yet i wonder if there is a use case where users **really** don't want to run the TokenCleaner on a cluster. :thinking: